File: rootfs/snapfaas/runtimes/go/workload.go
package main

import (
	"encoding/binary"
	"encoding/json"
	"os"
	"os/exec"
	"plugin"
	"runtime"

	"golang.org/x/sys/unix"
)

func main() {
	err := unix.Iopl(3)
	if err != nil {
		panic(err)
	}

	for i := 1; i < runtime.NumCPU(); i++ {
		exec.Command("taskset", "-c", string(i), "outl", "124", "0x3f0").Run()
	}
	exec.Command("taskset", "-c", "0", "outl", "124", "0x3f0").Run()

	if unix.Mount("/dev/vdb", "/srv", "ext2", unix.MS_RDONLY, "") != nil {
		panic("Failed to mount")
	}

	p, err := plugin.Open(os.Args[1])
	if err != nil {
		panic(err)
	}
	initptr, err := p.Lookup("Init")
	if err != nil {
		panic(err)
	}
	handleptr, err := p.Lookup("Handle")
	if err != nil {
		panic(err)
	}

	init := initptr.(func())
	handle := handleptr.(func([]byte) interface{})

	init()

	for i := 1; i < runtime.NumCPU(); i++ {
		exec.Command("taskset", "-c", string(i), "do-snapshot", "123").Run()
	}
	exec.Command("taskset", "-c", "0", "do-snapshot", "123").Run()

	fd, err := unix.Socket(unix.AF_VSOCK, unix.SOCK_STREAM, 0)
	if err != nil {
		panic(err)
	}
	sa := unix.SockaddrVM{
		CID:  2,
		Port: 1234,
	}
	err = unix.Connect(fd, &sa)
	if err != nil {
		panic(err)
	}

	for {
		lenbuf := make([]byte, 4)
		_, err = unix.Read(fd, lenbuf)
		if err != nil {
			panic(err)
		}
		reqlen := binary.BigEndian.Uint32(lenbuf)
		reqbuf := make([]byte, reqlen)
		_, err = unix.Read(fd, reqbuf)
		if err != nil {
			panic(err)
		}

		response := handle(reqbuf)

		responseBuf, err := json.Marshal(response)

		binary.BigEndian.PutUint32(lenbuf, uint32(len(responseBuf)))
		_, err = unix.Write(fd, lenbuf)
		if err != nil {
			panic(err)
		}
		unix.Write(fd, responseBuf)
	}
}

File: rootfs/snapfaas/runtimes/python3/workload.py
#!/usr/bin/env python3

from importlib import import_module
import os
import struct
import json
from subprocess import run, Popen
import time
import socket

import sys
import syscalls_pb2

# vsock to communicate with the host
VSOCKPORT = 1234
sock = socket.socket(socket.AF_VSOCK, socket.SOCK_STREAM)
hostaddr = (socket.VMADDR_CID_HOST, VSOCKPORT)

# for language snapshot
for i in range(1, os.cpu_count()):
    Popen('taskset -c {} outl 124 0x3f0'.format(i), shell=True)
run('taskset -c 0 outl 124 0x3f0', shell=True)

# mount appfs and load application
run(["mount", "-r", "/dev/vdb", "/srv"], executable="/bin/mount")
app = import_module('workload')


class Syscall():
    def __init__(self, sock):
        self.sock = sock

    def write_key(self, key, value):
        req = syscalls_pb2.Syscall(writeKey = syscalls_pb2.WriteKey(key = key, value = value))
        reqData = req.SerializeToString()
        self.sock.sendall(struct.pack(">I", len(reqData)))
        self.sock.sendall(reqData)

        data = sock.recv(4, socket.MSG_WAITALL)
        res = struct.unpack(">I", data)
        responseData = sock.recv(res[0], socket.MSG_WAITALL)

        response = syscalls_pb2.WriteKeyResponse()
        response.ParseFromString(responseData)
        return response.success

    def read_key(self, key):
        req = syscalls_pb2.Syscall(readKey = syscalls_pb2.ReadKey(key = key))
        reqData = req.SerializeToString()
        self.sock.sendall(struct.pack(">I", len(reqData)))
        self.sock.sendall(reqData)

        data = sock.recv(4, socket.MSG_WAITALL)
        res = struct.unpack(">I", data)
        responseData = sock.recv(res[0], socket.MSG_WAITALL)

        response = syscalls_pb2.ReadKeyResponse()
        response.ParseFromString(responseData)
        return response.value


# for function diff snapshot
for i in range(1, os.cpu_count()):
    Popen('taskset -c %d outl 124 0x3f0'%(i), shell=True)
run('taskset -c 0 outl 124 0x3f0', shell=True)

sock.connect(hostaddr)
while True:
    sc = Syscall(sock)

    data = sock.recv(4, socket.MSG_WAITALL)
    res = struct.unpack(">I", data)
    requestData = sock.recv(res[0], socket.MSG_WAITALL)

    request = syscalls_pb2.Request()
    request.ParseFromString(requestData)

    start = time.monotonic_ns()
    responseJson = {}
    # print("response: " + request.payload)

    try:
        temp = json.loads(request.payload)
        payload = temp['payload']
        print("the stripped down request: " + str(payload))
        responseJson = app.handle(payload)
    except:
        responseJson = { 'error': str(sys.exec_info()) }

    responseJson['duration'] = time.monotonic_ns() - start

    print(responseJson)

    response = syscalls_pb2.Syscall(response = syscalls_pb2.Response(payload = json.dumps(responseJson)))
    
    responseData = response.SerializeToString()

    sock.sendall(struct.pack(">I", len(responseData)))
    sock.sendall(responseData)
